x-pack/plugins/security_solution/server/endpoint/services/actions/clients/lib/normalized_external_connector_client.ts

```diff
@@ -0,0 +1,113 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+import type { IUnsecuredActionsClient, ActionsClient } from '@kbn/actions-plugin/server';
+import type { ActionTypeExecutorResult } from '@kbn/actions-plugin/common';
+import type { Logger } from '@kbn/logging';
+import type { ConnectorWithExtraFindData } from '@kbn/actions-plugin/server/application/connector/types';
+import { once } from 'lodash';
+import { stringify } from '../../../../utils/stringify';
+import { ResponseActionsClientError, ResponseActionsConnectorNotConfiguredError } from '../errors';
+
+export interface NormalizedExternalConnectorClientExecuteOptions<
+  TParams extends Record<string, any> = Record<string, any>,
+  TSubAction = unknown
+> {
+  params: {
+    subAction: TSubAction;
+    subActionParams: TParams;
+  };
+  spaceId?: string;
+}
+
+/**
+ * Handles setting up the usage of Stack Connectors for Response Actions and normalizes usage of
+ * Connector's Sub-Actions plugin between the `ActionsClient` and the `IUnsecuredActionsClient`
+ * client interfaces. It also provides better typing support.
+ */
+export class NormalizedExternalConnectorClient {
+  protected readonly getConnectorInstance: () => Promise<ConnectorWithExtraFindData>;
+
+  constructor(
+    protected readonly connectorTypeId: string,
+    protected readonly connectorsClient: ActionsClient | IUnsecuredActionsClient,
+    protected readonly log: Logger
+  ) {
+    this.getConnectorInstance = once(async () => {
+      let connectorList: ConnectorWithExtraFindData[] = [];
+
+      try {
+        connectorList = await this.getAll();
+      } catch (err) {
+        throw new ResponseActionsClientError(
+          `Unable to retrieve list of stack connectors in order to find one for [${this.connectorTypeId}]: ${err.message}`,
+          // failure here is likely due to Authz, but because we don't have a good way to determine that,
+          // the `statusCode` below is set to `400` instead of `401`.
+          400,
+          err
+        );
+      }
+      const connector = connectorList.find(({ actionTypeId, isDeprecated, isMissingSecrets }) => {
+        return actionTypeId === this.connectorTypeId && !isDeprecated && !isMissingSecrets;
+      });
+
+      if (!connector) {
+        this.log.debug(stringify(connectorList));
+        throw new ResponseActionsConnectorNotConfiguredError(this.connectorTypeId);
+      }
+
+      this.log.debug(
+        `Using [${this.connectorTypeId}] stack connector: "${connector.name}" (ID: ${connector.id})`
+      );
+
+      return connector;
+    });
+  }
+
+  private isUnsecuredActionsClient(
+    client: ActionsClient | IUnsecuredActionsClient
+  ): client is IUnsecuredActionsClient {
+    // The methods below only exist in the normal `ActionsClient`
+    return !('create' in client) && !('delete' in client) && !('update' in client);
+  }
```

Comment:
q: shouldn't this return `true` if *any* of the listed methods is missing?

Reply:
It could.... I only coded to ensure they are all `false` so that if they ever add any of these to the `IUnsecuredActionsClient` the check will continue to work

Reply:
I see, looks like a reasonable approach 👍